Metaspace improvements in QuarkusUnitTest (and dev mode!) - round 3

yrodiere opened the pull request:
Follows up on #35407 

Disclaimer: I didn't get the chance to run the whole test suite locally, so... fingers crossed :/

With these changes, I was able to run the Hibernate ORM quickstart, and after a few initial leaks (can't be avoided since we don't restart Vertx), new classloaders consistently got GC'd and I never got above 6 `QuarkusClassLoader` instances (believe me, that's not a lot for dev mode).

As to our own test suite... it's getting better, but we're far from being done. There are some annoying circular references caused by `io.netty.util.concurrent.GlobalEventExecutor#terminationFuture` and I think `io.quarkus.test.ClearCache#clearAnnotationCache` is not working as well as we want it to (or maybe should be used more often, e.g. in continuous testing). And there's probably more.

---

Regarding the content of this PR:

The first two commits are about improving the information we get when debugging or looking at heap dumps. Because believe me, when there's a dozen different "Augmentation Class Loader: TEST" in the same heap dump, it's not funny at all.

The next commit fixes a JDBC URL that seemed suspicious, but doesn't really impact anything (the test that uses this config triggers a startup failure on purpose before the database is started anyway).

~~The next commit change a few JDBC URLs to avoid a leak caused by H2 when it registerers a JVM Shutdown Hook (which lasts forever) which indirectly references a `QuarkusClassLoader` (which is supposed to be short-lived).~~ => Removed, see https://github.com/quarkusio/quarkus/pull/36560#issuecomment-1771254657

Then come a few commits that impact dev mode, so might have an impact in the real world out there. Because it turns out a lot of the leaks in our test suite are actually caused by tests of dev mode, since dev mode is _very_ leaky: basically each dev mode restart leaks a whole application worth of classloaders.

These last commits, in order:

* ~~We apply the same fix for H2 JVM Shutdown Hooks, but this time for dev services~~ => Removed, see https://github.com/quarkusio/quarkus/pull/36560#issuecomment-1771254657
* We make sure to reset the TCCL of `ForkJoinPool#commonPool` threads for *every* `CuratedApplication#run` method (for some reason there are 3 of them)
* We simplify Smallrye config setting/release so that we don't end up with a classloader high in the hierarchy referencing a version of `io.smallrye.config.SmallRyeConfigProviderResolver` that will forever reference a classloader lower in the hierarchy, making it impossible to GC. cc @mkouba : this is the problem we've been investigating the other day; I tried a few solutions and this one seems to be both safe and effective.
* And finally, we fix a very extensive leak that probably affects every application in dev mode: `java.lang.Thread#inheritedAccessControlContext` has an array of `ProtectionDomain` with references to classloaders. In effect, any time we create a thread, it has a high chance of preventing several classloaders from being GC. This is very annoying for Vertx threads and `ForkJoinPool#commonPool` threads, in particular. cc @Sanne:, who noticed this independently, but much sooner than I did :)

yrodiere committed:
Safer, slightly simpler register/release patterns for config

Some config was lazily created and bound to the dev class loader
through the configsForClassLoader map
io.smallrye.config.SmallRyeConfigProviderResolver#getConfig(java.lang.ClassLoader),
and the corresponding map entry was never cleared,
resulting in the corresponding classloader never being garbage
collected.

yrodiere committed:
Enrich classloader names with the name of the application/QuarkusUnitTest

yrodiere committed:
Always reset the ForkJoinPool TCCL on startup

yrodiere committed:
Avoid referencing ClassLoaders in ProtectionDomains

This removes a whole class of metaspace leaks caused by
Thread.inheritedAccessControlContext referencing ProtectionDomains which
reference older classloaders.

Of course this may have impacts on how the SecurityManager behaves, but
as I understand it, absolutely no part of Quarkus is ready to run with a
SecurityManager enabled anyway.

yrodiere committed:
Clarify why QuarkusClassLoader doesn't pass its name to the super constructor

yrodiere commented:
Ok, now I understand why `QuarkusClassLoader` doesn't pass its name to the super constructor: to avoid ugly stacktraces. I'll change the first commit.

gsmet commented:
@radcortez could you have a look at the commit related to config?

gsmet reviewed (commented):
Added a small comment.

The changes looks good to me but I would like to have @radcortez to have a closer look at the config commit.

gsmet commented:
Also, @stuartwdouglas could you have a look at the `ProtectionDomain` commit, just to make sure we haven't missed something?

yrodiere commented on test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusUnitTest.java:
```diff
@@ -636,7 +636,8 @@ public boolean test(String s) {
                     quarkusUnitTestClassLoader = QuarkusClassLoader
-                            .builder("QuarkusUnitTest ClassLoader for " + extensionContext.getDisplayName(), getClass().getClassLoader(), false)
+                            .builder("QuarkusUnitTest ClassLoader for " + extensionContext.getDisplayName(),
+                                    getClass().getClassLoader(), false)
```
Right, I tried to squash but failed. Will force-push.

Sanne commented:
+1 for the H2 property. We're using the same in Hibernate ORM - for the same reasons.

> Also, @stuartwdouglas could you have a look at the `ProtectionDomain` commit, just to make sure we haven't missed something?

This should be fine, as mentioned elsewhere we never designed Quarkus to be used with the SecurityManager. In fact I suggested this same change to Yoann and plan to apply the same to the other classloaders, with some additional optimisations.

Sanne reviewed (approved):
I approve, pending Roberto's opinion as suggested by Guillaume.

stuartwdouglas commented:
I know that DB_CLOSE_DELAY=-1 has bit us before with race conditions in tests. Basically if the number of connections drops to zero then the DB gets deleted in the middle of the test. This was a long time ago that we were seeing this so I don't remember all the details.

stuartwdouglas commented:
Actually ignore that last comment, I misremembered what the different options do, and this looks fine.

stuartwdouglas commented:
I think the ProtectionDomain changes are fine.

radcortez commented on core/runtime/src/main/java/io/quarkus/runtime/configuration/QuarkusConfigFactory.java:
```diff
@@ -19,21 +19,37 @@ public QuarkusConfigFactory() {
-        QuarkusConfigFactory.config = config;
+        // Uninstall previous config
         if (QuarkusConfigFactory.config != null) {
+            configProviderResolver.releaseConfig(QuarkusConfigFactory.config);
```
While unlikely, it is possible to release / register the Config outside this Factory, invalidating the cached referenced when calling `getConfigFor` for `SmallRyeConfigProviderResolver`

It is safer to just release whatever Config is registered to the TCCL or a specific CL (that is why we added that the release method per CL).

Is there a reason for going back to release per Config and not per CL?

yrodiere replied:
> While unlikely, it is possible to release / register the Config outside this Factory, invalidating the cached referenced when calling `getConfigFor` for `SmallRyeConfigProviderResolver`

If someone calls `SmallRyeConfigProviderResolver` release/register, I'm going to guess that invalidating the cached reference in `SmallRyeConfigProviderResolver` is precisely what they want, so I don't see the problem?

> Is there a reason for going back to release per Config and not per CL?

Yes.

`QuarkusConfigFactory#getConfigFor` actually provides a default config for whenever `SmallRyeConfigProviderResolver` is asked for the config of a classloader and none is registered yet. As a result, there are many more classloaders that can end up bound to `QuarkusConfigFactory.config` than just the TCCL when we call `setConfig`.

How it works: when someone retrieves the config via `SmallRyeConfigProviderResolver` for a given classloader, and that classloader is the not the TCCL that was active when calling `setConfig`,  `SmallRyeConfigProviderResolver` will call `QuarkusConfigFactory.getConfigFor`, and ultimately insert an entry in `SmallRyeConfigProviderResolver#configsForClassLoader` with that classloader as key and the config as value .

If we only release config for the TCCL, this entry for another classloader will effectively persist forever, and so will the reference to the classloader, and you got yourself a metaspace leak.
That's why I call `releaseConfig(config)`: to invalidate the cache of *all* classloaders that used this config, and avoid leaks.

Note that we *also* release per-CL where it makes sense: see `releaseTCCLConfig` and how it's called in `ConfigRecorder`. I changed the way we release because the existing calls to `setConfig(null)` were actually using a `QuarkusConfigFactory` and a `SmallRyeConfigProviderResolver` from a different classloader than the initial call to `setConfig(<non-null>)` in the generated config class.

Anyway... it's way too complex for its own good, and that's why I wanted to make Smallrye Config parent-first, but that didn't fly because people had performance concerns. I tried many alternatives before settling for this one, and most were just breaking the tests. This one fixes the leaks **and** doesn't break the tests.

yrodiere replied:
Alright, fair point, I did not know that `registerConfig` failed when there was already a registered config. Thanks for pointing that out.

I amended the commit to add a `releaseConfig(TCCL)` as you suggested. Is this better?

radcortez replied:
Yes, thank you :)

yrodiere commented:
> Actually ignore that last comment, I misremembered what the different options do, and this looks fine.

Looks like it used to be fine... but the upgrade to H2 2.2.224 (merged while I was working on this) changes things. The behavior was altered [here](https://github.com/h2database/h2database/commit/f62de75a5e4a46d17af6010c17236ab87eae2d05#diff-381f07093dd2ad7c477f3009b3f9eebf13864ccb401c440288546ff7174961d8R563-R566) and now we have failing tests.

I will have a look again...

yrodiere committed:
Fix a suspicious H2 URL in Hibernate ORM test config

It doesn't matter in practice since the only test using this config
reproduces a startup failure that happens before we start H2;
but still, let's fix this for correctness.

yrodiere commented:
> > Actually ignore that last comment, I misremembered what the different options do, and this looks fine.
> 
> Looks like it used to be fine... but the upgrade to H2 2.2.224 (merged while I was working on this) changes things. The behavior was altered [here](https://github.com/h2database/h2database/commit/f62de75a5e4a46d17af6010c17236ab87eae2d05#diff-381f07093dd2ad7c477f3009b3f9eebf13864ccb401c440288546ff7174961d8R563-R566) and now we have failing tests.
> 
> I will have a look again...

Alright, using DB_CLOSE_ON_EXIT=false is no longer necessary with H2 2.2.224. If the DB is not persistent (in-memory), the H2 2.2.224 won't register the shutdown hook anyway, which is exactly what we want.

So DB_CLOSE_ON_EXIT didn't cause the tests to fail after all; the various DB_CLOSE_DELAY=-1 that I added did. They cause the same DB content to be reused across tests, which is definitely not something that we want.

I'll remove all the H2-related stuff from this PR, since it's no longer necessary and (in the case of DB_CLOSE_DELAY=-1) potentially dangerous.

FWIW I'm not sure DB_CLOSE_DELAY=-1 is ever necessary with H2 2.x... As far as I can see the default of `0` should close the DB as soon as all connections are closed, which should be what we want in most cases (tests, production); the only case it may not be what we want is in dev mode, if we restart Agroal but don't want to restart the DB.
Anyway... I'll leave that battle to someone else.

cc @Sanne ^

yrodiere committed:
Improve error messages on KafkaDevServicesDevModeTestCase failures

yrodiere commented:
Thanks all.

Unfortunately the `KafkaDevServicesDevModeTestCase` is legitimate. I'm investigating.

yrodiere committed:
Plug leak in tests relying on lazily-initialized config

Since a previous commit, we now remember the lazily-created config in
QuarkusConfigFactory, which is good because in dev mode it allows us to
release it upon restart (when we call setConfig(null)).

However, in *tests* that rely on this lazily initialized config, just
calling releaseConfig(ConfigProvider.getConfig()) is no longer enough,
because of that remembered config in QuarkusConfigFactory: we need to
reset that reference too.

If we forget to reset it, then when KafkaDevServicesDevModeTestCase
will execute, TestHTTPResourceManager#getUri will retrieve the leaked
config from DefaultSerdeConfigTest or DefaultSchemaConfigTest,
the injected URI in KafkaDevServicesDevModeTestCase will be wrong,
and the test will fail (with very unhelpful error messages).

yrodiere committed:
More comprehensive registration/release of config in some tests

If someone calls ConfigProvider.getConfig() out of the blue in a test
that doesn't use any Quarkus*Test extension, this will indirectly call
QuarkusConfigFactory and leak config in two ways:

1. In QuarkusConfigFactory#config
2. In SmallRyeConfigProviderResolver, registering config for the TCCL,
   which in such a case is most likely the system CL.

Thus, a well-behaved test should call QuarkusConfigFactory.setConfig(null)
to clean up all that, no just SmallRyeConfigProviderResolver.releaseConfig().

Similarly, tests that register configuration explicitly can just call
QuarkusConfigFactory.setConfig(config) at the beginning and
QuarkusConfigFactory.setConfig(null) at the end,
which will properly simulate how a real Quarkus application behaves,
and should cover all edge cases involving multiple classloaders,
properly cleaning up everything at the end of the test.

yrodiere committed:
Stronger safeguards against leaked config in internal Quarkus*Test extensions

If someone calls ConfigProvider.getConfig() out of the blue in a test
that doesn't use any Quarkus*Test extension, this will call
QuarkusConfigFactory and leak config in two ways:

1. In QuarkusConfigFactory#config
2. In SmallRyeConfigProviderResolver, registering config for the TCCL,
   which in such a case is most likely the system CL.

A well-behaved test would call QuarkusConfigFactory.setConfig(null) to
clean up all that, but it's easy to miss and there is potential for
ConfigProvider.getConfig() being called indirectly, so there's no way we
can guarantee all tests are well-behaved.

This should at least guarantee that after a badly behaving test
executes, the next test using a Quarkus*Test extension will clean up the
mess.

yrodiere commented:
> Unfortunately the `KafkaDevServicesDevModeTestCase` failure is legitimate. I'm investigating.

Turns out another test that doesn't use any `Quarkus*Test` extension was leaking its config, leading to `TestHTTPResourceManager#getUri` injecting a wrong URI into `KafkaDevServicesDevModeTestCase`.

Hey @radcortez, do you think you could have a look at the last three commits I just added? The commit messages explain what they are about. Only the first one is strictly necessary, but I think to avoid other such problems in the future we'd better merge the other two as well.

I didn't alter the "external" test extensions (`QuarkusTestExtension`, `QuarkusIntegrationTestExtension`) on purpose, because I didn't want to cause accidental regressions there (some test use `QuarkusDevModeTest` and within it they do continuous testing with `QuarkusTestExtension`, so resetting things in `QuarkusTestExtension` is a bit tricky).

radcortez commented on extensions/config-yaml/runtime/src/test/java/io/quarkus/config/yaml/runtime/ApplicationYamlTest.java:
```diff
@@ -21,19 +20,12 @@
-            cpr.releaseConfig(conf);
-            ConfigProvider.getConfig();
-        }
+        SmallRyeConfig config = builder.build();
```
This can be simplified. If we are getting a `SmallRyeConfig` instance from the builder and testing against it, we don't need to register it into the factory, because this is not even a Quarkus test.

radcortez commented on extensions/resteasy-classic/rest-client-config/runtime/src/test/java/io/quarkus/restclient/config/RestClientConfigTest.java:
```diff
@@ -8,29 +8,28 @@
+import io.smallrye.config.SmallRyeConfigBuilder;
+import io.smallrye.config.SmallRyeConfigProviderResolver;
 
 public class RestClientConfigTest {
```
We can simplify this test by using `SmallRyeConfigBuilder` and `SmallRyeConfig`, and we don't need to set / release stuff, because this is not a Quarkus test.

radcortez commented on test-framework/junit5-internal/src/main/java/io/quarkus/test/ConfigUtil.java:
```diff
@@ -0,0 +1,30 @@
+
+import io.quarkus.runtime.configuration.QuarkusConfigFactory;
+
+public final class ConfigUtil {
```
Not another `ConfigUtil`... we already have many of these (or similar). I know this will be for test only, and this method shouldn't go the main one, but I already get frustrated when I search for it to get 3 or 4 alternatives and always pick the wrong one :)

Maybe this could go into `io.quarkus.test.junit.launcher.ConfigUtil`?

yrodiere replied:
Unfortunately I don't have access to `io.quarkus.test.junit.launcher.ConfigUtil`  from `quarkus-junit5-internal`, because it's in its sibling module `quarkus-junit5`...

I'll merge the two `ConfigUtils` and move the result to `quarkus-test-common`, with a more unique name (`TestConfigUtil`). If you think `io.quarkus.test.junit.launcher.ConfigUtil` is API, I can leave it in place and deprecate it; just let me know.

radcortez replied:
I don't think it is an API class. Feel free to merge and move :)

radcortez reviewed (changes requested):
I'm not a big fan of cleaning up mess like this, but I guess it is easier this way.

My recommendation is for tests that are not QuarkusTests to just use `SmallRyeConfigBuilder` and `SmallRyeConfig` directly instead of `CofigProvider.getConfig`.

yrodiere committed:
Merge test-related ConfigUtils into a single class

yrodiere commented:
Thanks for the review.

> I'm not a big fan of cleaning up mess like this, but I guess it is easier this way.

Right, it's more of a safety net. One that's apparently necessary as I saw several similar cleanups throughout the test modules.
Ideally we'd just have tests that don't leak, but I don't think anyone wants to deal with that right now.

Anyway... I addressed your comments and switched this PR to draft. Now I'll have a look at the test failures... This time I won't mark as ready for review until my fork's build passes.

yrodiere commented:
Alright the build passed on my fork (if we ignore some flakes). 

@radcortez I had to revert a change to `TestResourceManager`; apparently using QuarkusTestFactory there does more harm than good. I'm not sure even using `releaseConfig` there makes sense now, but that at least doesn't seem to hurt so I left it and added a comment: https://github.com/quarkusio/quarkus/pull/36560/commits/e76a0e537f4588fa268cf58058ff4dc1a764560e#diff-a050ea6d2087e6491a6cfcadd0ef58e0ffbe09935ab12850ca3adbc38ce14bffR182-R188

yrodiere commented:
From what I can see, the failing builds are failing on other PRs as well:

* `Native Tests - Virtual Thread - Main`
  * https://github.com/quarkusio/quarkus/pull/36753
  * https://github.com/quarkusio/quarkus/pull/36732
* `Native Tests - Windows - RESTEasy Jackson`
  * https://github.com/quarkusio/quarkus/pull/36771
  * https://github.com/quarkusio/quarkus/pull/36753

I looked at the logs and they don't seem to point at any problem that would obviously come from from this PR. `JVM Tests - JDK 17` and `JVM Tests - JDK 17 Windows`, in particular, just took more time than usual for no obvious reason.

So, I'll merge. Thanks for the reviews!
